samples/snippets/src/main/java/pubsub/SubscribeWithAvroSchemaExample.java

```diff
@@ -0,0 +1,123 @@
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.TopicName;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.specific.SpecificData;
+import utilities.State;
+
+public class SubscribeWithAvroSchemaExample {
+  public static void main(String... args) throws Exception {
+    // TODO(developer): Replace these variables before running the sample.
+    String projectId = "your-project-id";
+    // Use a topic that has a schema attached.
+    String topicId = "your-topic-id";
+    // Use a subscription attached to your topic.
+    String subscriptionId = "your-subscription-id";
+
+    subscribeWithAvroSchemaExample(projectId, topicId, subscriptionId);
+  }
+
+  public static void subscribeWithAvroSchemaExample(String projectId, String topicId, String subscriptionId)
+      throws IOException {
+
+    TopicName topicName = TopicName.of(projectId, topicId);
+    ProjectSubscriptionName subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId);
+
+    // Get the topic encoding type.
+    Encoding encoding = null;
+    try (TopicAdminClient topicAdminClient = TopicAdminClient.create()) {
+      encoding = topicAdminClient.getTopic(topicName).getSchemaSettings().getEncoding();
+    }
+    Encoding finalEncoding = encoding;
+
+    // Prepare a reader for incoming Avro records.
+    GenericDatumReader<GenericRecord> reader = new GenericDatumReader<>(
+        State.getClassSchema()
+    );
+
+    // Instantiate an asynchronous message receiver.
+    MessageReceiver receiver = (PubsubMessage message, AckReplyConsumer consumer) -> {
+
+      ByteString data = message.getData();
+
+      // Send the message data to a byte[] input stream.
+      InputStream inputStream = new ByteArrayInputStream(data.toByteArray());
+
+      Decoder decoder = null;
+
+      // Prepare an appropriate decoder for messages in this subscription.
+      block: try {
+        switch (finalEncoding) {
+          case BINARY:
+            decoder = DecoderFactory.get().directBinaryDecoder(inputStream, /*reuse=*/null);
+            System.out.println("Receiving a binary-encoded message:");
+            break;
+          case ENCODING_UNSPECIFIED:
+          case UNRECOGNIZED:
+            break block;
+          case JSON:
+            decoder = DecoderFactory.get().jsonDecoder(State.getClassSchema(), inputStream);
+            System.out.println("Receiving a JSON-encoded message:");
+            break;
+        }
+
+        // Obtain a generic Avro record using the decoder.
+        GenericRecord record = reader.read(null, decoder);
```

Comment:
I don't think it's necessary to go through the GenericRecord object, is it? Could you create a SpecificDatumReader<State> that reads out a state object?

I think you'd have something like:

SpecificDatumReader<State> reader = new SpecificDatumReader<>(State.getClassSchema());
...
State state = reader.read(null, decoder);

Reply:
Done. I had a mistake in my .avsc file earlier that caused some errors with SpecificDatumReader.